### PR TITLE
feat: switch to react-component

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Switch to React component:
- remove code related to Nunjucks (remove something like 90% of template 😅).
- unfortunately template is still written in Nunjucks, not in React, due to https://github.com/asyncapi/generator/issues/521 issue. After fixing it, rewriting ro React will be very easy.
- write `scripts/copy-sources.js` script to copy to template's files all needed sources to use template:
  - react-dom
  - react
  - asyncapi-react component
  - default (also minified) styles for react-component
- update Readme.md
- write needed filters and tests for them
- update hooks

The easiest way to test template from PR is run generator by:

```sh
ag https://gist.githubusercontent.com/derberg/4e419d6ff5870c7c3f5f443e8bd30535/raw/6188903caf77ad14ee32e22b0c86bf6fa244b580/asyncapi-websocket-kraken.yml https://raw.githubusercontent.com/asyncapi/html-template/989c2ba26f228e030550dea285bb09bcdad92b57/asyncapi-html-template-0.20.1.tgz -o ./sample --force-write
``` 

Don't afraid 😅 I didn't put any virus to `.tgz`. You can check that this package is inside the PR. Of course I will remove it before merging to master. 

**Related issue(s)**
Part of https://github.com/asyncapi/shape-up-process/issues/88

Blocked by https://github.com/asyncapi/shape-up-process/issues/86
Blocked by https://github.com/asyncapi/asyncapi-react/pull/301
Blocked by https://github.com/asyncapi/asyncapi-react/pull/302

Fixes https://github.com/asyncapi/html-template/issues/168
Fixes https://github.com/asyncapi/html-template/issues/167
Fixes https://github.com/asyncapi/html-template/issues/15
Closes https://github.com/asyncapi/html-template/issues/160
Closes https://github.com/asyncapi/html-template/issues/51

### CODE REVIEW

File: filters/all.js
Comment (fmvilas) [inline filters/all.js:1]: I see you're using Nunjucks filters. I'd avoid using them because we're gonna get rid of Nunjucks sooner than later. Maybe you should place all the logic in a `generate:after` hook and convert this template into a React template?

File: template/index.html
Comment (fmvilas) [inline template/index.html:3]: Same as above, you should probably convert this template into a React template.

**Review by fmvilas — CHANGES_REQUESTED** [review]
Exciting to see this coming! 😄

File: filters/all.js
Comment (magicmatatjahu) [reply filters/all.js:1]: You can see in commits that I had a template written in React, but due to this https://github.com/asyncapi/generator/issues/521 I wanna stay with Nunjucks. When mentioned issue will be fixed, rewritten to React will be very easy.

File: template/index.html
Comment (magicmatatjahu) [reply template/index.html:3]: You can see in commits that I had a template written in React, but due to this asyncapi/generator#521 I wanna stay with Nunjucks. When mentioned issue will be fixed, rewritten to React will be very easy :)

File: README.md
Comment (derberg) [inline README.md:12]: would be great to have weekly or monthly downloads badge 😄

**Review by derberg — CHANGES_REQUESTED** [review]
- contribution flow will not be super friendly now, for now please at least make sure react component has https://github.com/asyncapi/generator/blob/master/.github/workflows/bump.yml flow to update dependants automatically
- we need a PR for playground too
- tests are removed cause filters are removed, but I see new functions are added, but not tested

File: README.md
Comment (magicmatatjahu) [reply README.md:12]: I wrote year when Kyma give AsyncAPI 100% control of project :) I don't know if we need to clarify it.

File: filters/all.js
Comment (derberg) [reply filters/all.js:2]: then why not fixing https://github.com/asyncapi/generator/issues/521 first?

Comment (magicmatatjahu) [reply filters/all.js:2]: I don't want to end the cycle without merging neither the component nor the html-template like in previous cycle 😄 Taks with bundler took a lot longer than I thought and the current PR would have opened in the past week. I want to deal with this task, but after done the current tasks.

File: README.md
Comment (magicmatatjahu) [reply README.md:12]: I added monthly :)

**Comment by magicmatatjahu** [comment]
@derberg Thanks!

I added tests for new filters.

> contribution flow will not be super friendly now, for now please at least make sure react component has https://github.com/asyncapi/generator/blob/master/.github/workflows/bump.yml flow to update dependants automatically

Good catch! I will add it, because component hasn't it.

> we need a PR for playground too

I have PR https://github.com/asyncapi/asyncapi-react/pull/301 in React component with styling vol.2 (for examples and sidebar) and probably only one possibility is fetch my fork with my branch and run playground locally. A nice solution to connect Netlify to our playground, I will create issue for that.

**Comment by magicmatatjahu** [comment]
@derberg Issue is created https://github.com/asyncapi/asyncapi-react/issues/306

**Comment by fmvilas** [comment]
Code-wise, it looks great to me. I'm only seeing a problem with the expand/collapse buttons not being clickable (only the chevron icon is). Here's a video of what I mean: https://www.loom.com/share/3b7037ddd3074796a4774dbfe52790e4

**Comment by magicmatatjahu** [comment]
@fmvilas Thanks for review! :) I left one PR in react-component repo for such bugs. I will correct it! :)

**Comment by magicmatatjahu** [comment]
@fmvilas Fixed! You can check it by command:

```bash
ag https://raw.githubusercontent.com/asyncapi/generator/master/test/docs/dummy.yml https://raw.githubusercontent.com/asyncapi/html-template/6714302c5276bd3b7c575e23bb1445d83312bdc3/asyncapi-html-template-0.20.1.tgz -o ./sample --force-write -p version='1.2.5' -p sidebarOrganization=byTagsNoRoot -p singleFile=true
```

Here is a new tgz:

```bash
https://raw.githubusercontent.com/asyncapi/html-template/6714302c5276bd3b7c575e23bb1445d83312bdc3/asyncapi-html-template-0.20.1.tgz
```

**Review by derberg — CHANGES_REQUESTED** [review]
> I have PR asyncapi/asyncapi-react#301 in React component with styling vol.2 (for examples and sidebar) and probably only one possibility is fetch my fork with my branch and run playground locally. A nice solution will be connect Netlify to our playground, I will create issue for that.

There seems to be a misunderstanding here 😄 I mean that once this is released, you need to manually update it here https://playground.asyncapi.io/ and make sure it works


---
I did manual testing with:

- https://bitbucket.org/qbmt/zbos-mqtt-api/raw/25ee18cee0db90d9f370e0bcf1bfbb6fbb7621ce/zbos_mqtt-all-asyncapi.json and when you have loads of messages (over 160) then expanding collapsable elements have delay, but yeah, not a blocker but something to think about in future
- https://gist.githubusercontent.com/derberg/4e419d6ff5870c7c3f5f443e8bd30535/raw/6188903caf77ad14ee32e22b0c86bf6fa244b580/asyncapi-websocket-kraken.yml has issue with container containing description that is super narrow + problem with rendering examples
Screenshot 2021-04-22 at 13 22 51
Screenshot 2021-04-22 at 13 22 24

also this `anonymous-message-1` should not be there 🤔 maybe my messages do not have `name` prop or `title` but they are in components and the key name should be applied right?
Screenshot 2021-04-22 at 13 32 58


---
overall though, super great job mate, I love that when there are objects in extensions and bindings, you do not just dump but try rendering in a nice way ❤️

**Comment by magicmatatjahu** [comment]
@derberg Thanks for review man! 

> overall though, super great job mate, I love that when there are objects in extensions and bindings, you do not just dump but try rendering in a nice way ❤️

🍻 

I will address all bugs what you found, thanks you again for that! Some are very strange like `description is super narrow`... Probably I missed about one css class when I was splitting [this PR](https://github.com/asyncapi/asyncapi-react/pull/301).

> https://bitbucket.org/qbmt/zbos-mqtt-api/raw/25ee18cee0db90d9f370e0bcf1bfbb6fbb7621ce/zbos_mqtt-all-asyncapi.json and when you have loads of messages (over 160) then expanding collapsable elements have delay, but yeah, not a blocker but something to think about in future

Hmmm... I tried write collapse button in this way that shouldn't be rendered in each collapse... Examples haven't this delay 🤔 
But yeah, I will create issue for that. 

> also this anonymous-message-1 should not be there 🤔 maybe my messages do not have name prop or title but they are in components and the key name should be applied right?

Strange, I will check that!

**Comment by fmvilas** [comment]
Lovely! Yeah, I agree with @derberg. The only part that can be more challenging is taking the key on the components section as the name of the message because by the time you're processing the document it's already dereferenced and you don't have the context. Anyway, that's a problem to solve in the parser too, not a blocker for me.

**Comment by derberg** [comment]
I noticed it only because in the current playground I don't see those anonymous tags

**Comment by magicmatatjahu** [comment]
I checked this Kraken API in my local playground with newest component and it renders like:

image

Probably there is a problem with hydration.

**Comment by magicmatatjahu** [comment]
@derberg Problems with

- delay in collapse buttons
- narrow markdowns
- invalid render of examples in extensions

is fixed in

```bash
ag https://bitbucket.org/qbmt/zbos-mqtt-api/raw/25ee18cee0db90d9f370e0bcf1bfbb6fbb7621ce/zbos_mqtt-all-asyncapi.json https://raw.githubusercontent.com/asyncapi/html-template/56af23997f5848e65f7344b3a24b515ed26bc70d/asyncapi-html-template-0.20.1.tgz -o ./sample --force-write
```

The problem with `anonymous-messages` are described and fixed in this PR in parser (this is related not to component but to parser) - https://github.com/asyncapi/parser-js/pull/288

Additionally, I would not like to release such a big changes in Friday. I would prefer to wait until Monday to be on guard as soon as there are problems. Any objections?

**Comment by derberg** [comment]
> Additionally, I would not like to release such a big changes in Friday. I would prefer to wait until Monday to be on guard as soon as there are problems. Any objections?

**Comment by derberg** [comment]
I noticed 2 bugs with custom params:
- baseHref breaks UI rendering
- single file generation fails:
```bash
ag https://bitbucket.org/qbmt/zbos-mqtt-api/raw/25ee18cee0db90d9f370e0bcf1bfbb6fbb7621ce/zbos_mqtt-all-asyncapi.json https://raw.githubusercontent.com/asyncapi/html-template/56af23997f5848e65f7344b3a24b515ed26bc70d/asyncapi-html-template-0.20.1.tgz -o ./sample --force-write -p singleFile=true
Something went wrong:
Template render error: (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/node_modules/@asyncapi/html-template/template/index.html) [Line 7, Column 57]
  Error: ENOENT: no such file or directory, open 'template/css/styles.min.css'
    at Object._prettifyError (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/node_modules/nunjucks/src/lib.js:36:11)
    at /Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/node_modules/nunjucks/src/environment.js:563:19
    at Template.root [as rootRenderFunc] (eval at _compile (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/node_modules/nunjucks/src/environment.js:633:18), <anonymous>:58:3)
    at Template.render (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/node_modules/nunjucks/src/environment.js:552:10)
    at Environment.renderString (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/node_modules/nunjucks/src/environment.js:380:17)
    at /Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/renderer/nunjucks.js:32:14
    at new Promise (<anonymous>)
    at nunjucksExport.renderNunjucks (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/renderer/nunjucks.js:31:10)
    at Generator.renderFile (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/generator.js:688:12)
    at Generator.renderAndWriteToFile (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/generator.js:602:27)
    at Generator.generateFile (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/generator.js:652:5)
    at Generator.filesGenerationHandler (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/generator.js:527:7)
    at Walker.<anonymous> (/Users/wookiee/.nvm/versions/node/v12.18.3/lib/node_modules/@asyncapi/generator/lib/generator.js:451:11)
```

**Comment by magicmatatjahu** [comment]
@derberg Thanks again for review! 🙇🏼‍♂️ 

> baseHref breaks UI rendering

You probably pass `/docs` or something like it and go to generated html in your machine, so then you will have paths like `file:///docs/...`. You can also check `html-template@0.20.1` and result will be the same.

> single file generation fails:

🤦🏼 My fault! I changed to absolute path.

Here is (probably last) tgz to test, also with fixed `anonymous-message` bug :)

```bash
https://raw.githubusercontent.com/asyncapi/html-template/989c2ba26f228e030550dea285bb09bcdad92b57/asyncapi-html-template-0.20.1.tgz
```

```bash
ag https://gist.githubusercontent.com/derberg/4e419d6ff5870c7c3f5f443e8bd30535/raw/6188903caf77ad14ee32e22b0c86bf6fa244b580/asyncapi-websocket-kraken.yml https://raw.githubusercontent.com/asyncapi/html-template/989c2ba26f228e030550dea285bb09bcdad92b57/asyncapi-html-template-0.20.1.tgz -o ./sample --force-write -p version='1.2.5' -p sidebarOrganization=byTagsNoRoot -p singleFile=true
```

**Review by fmvilas — APPROVED** [review]
Looks good to me. If fixing the `baseHref` param problem is not a lot of work, it would be great to have it fixed before merging. Otherwise, just create a follow-up issue.

**Comment by magicmatatjahu** [comment]
> Looks good to me. If fixing the baseHref param problem is not a lot of work, it would be great to have it fixed before merging. Otherwise, just create a follow-up issue.

I will create issue, because it's not a trivial problem. It also occurs in currently html-template and as I read in internet, handling anchors with baseHref is (probably) only possible with JS, but I don't want to do this in this time.

**Comment by magicmatatjahu** [comment]
@derberg Created PR with playground https://github.com/asyncapi/playground/pull/117 :)

**Review by derberg — DISMISSED** [review]

**Comment by magicmatatjahu** [comment]
@derberg @fmvilas I updated dependency of `react-component` to `1.0.0-next.1` :) Here is a tgz to test:

```text
https://raw.githubusercontent.com/asyncapi/html-template/2aff6fba4732c88d414aa05964599a70b28793a4/asyncapi-html-template-0.20.1.tgz
```

```
ag https://gist.githubusercontent.com/derberg/4e419d6ff5870c7c3f5f443e8bd30535/raw/6188903caf77ad14ee32e22b0c86bf6fa244b580/asyncapi-websocket-kraken.yml https://raw.githubusercontent.com/asyncapi/html-template/2aff6fba4732c88d414aa05964599a70b28793a4/asyncapi-html-template-0.20.1.tgz -o ./sample --force-write -p version='1.2.5' -p sidebarOrganization=byTagsNoRoot -p singleFile=true
```

**Review by derberg — COMMENTED** [review]
should this one be `next` release too? in the end it now relies fully on react component, structure of generated files changes, and when react reaches 1.0, then this template will reach 1.0 too, right?

open question here

**Comment by derberg** [comment]
also separate topic, but I guess we should consider moving this repo as part of react component monorepo?

**Comment by magicmatatjahu** [comment]
@derberg With `next` releases, it is a good question. I'm also open here. @fmvilas What do you think? 

> also separate topic, but I guess we should consider moving this repo as part of react component monorepo?

I had a lot of problems with testing react-component in html-template to have 100% sure that everything works, I mean making symlinks, sometimes symlinks were broken, so I had to reinstall deps etc... so I like your idea :) But we must ask the core contributors to this repo, what about it they think.

**Comment by derberg** [comment]
> I had a lot of problems with testing react-component in html-template to have 100% sure that everything works, I mean making symlinks, sometimes symlinks were broken, so I had to reinstall deps etc... so I like your idea :) But we must ask the core contributors to this repo, what about it they think.

yeah, I think any maintainer would prefer to have it in one repo instead, but for now forget about it, I forgot generator would not be able to use such template, it doesn't support monorepo

**Comment by magicmatatjahu** [comment]
> yeah, I think any maintainer would prefer to have it in one repo instead, but for now forget about it, I forgot generator would not be able to use such template, it doesn't support monorepo

We can make similar approach like in web-component so user will be able to use template like now by `@asyncapi/html-template`. What is a problem here? Maybe I don't know about something 😅

**Comment by derberg** [comment]
@magicmatatjahu I quite often install template from someones fork, or mine fork using something like `https://github.com/USERNAME_WHERE_FORK_IS/html-template#BRANCH_NAME`

**Comment by magicmatatjahu** [comment]
@fmvilas What idea do you have? Together with Łukasz, we are open to suggestions. Go with 1.0.0-next or publish html as 0.21.0?

**Comment by fmvilas** [comment]
I have no preference, to be honest. Either way would be fine for me.

**Comment by jonaslagoni** [comment]
Same as Fran, no preference here, especially since the template is < v1.

**Review by derberg — APPROVED** [review]

**Comment by magicmatatjahu** [comment]
@all-contributors please add @magicmatatjahu for code, test, bug

**Pull request merged**
